### PR TITLE
Debug health check

### PR DESCRIPTION
**Pull request by azhavoro**
### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into [CHANGELOG](https://github.com/cvat-ai/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/cvat-ai/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

**Comment by nmanovic** [comment]
@azhavoro , can we customize the error message in the PR?

File: cvat-ui/src/utils/enviroment.ts [hunk 1]
```diff
@@ -29,3 +29,7 @@ export function customWaViewHit(pageName?: string, queryString?: string, hashInf
         }
     }
 }
+
+export function isPublicInstance(): boolean {
```
Comment (nmanovic) [inline cvat-ui/src/utils/enviroment.ts:33]: I don't think that it is a right solution. Probably HealthCheck endpoint can return extra information. Need to think how to implement that. Another solution is to have the message as a variable in settings. Please think and propose the best solution.

**Comment by nmanovic** [comment]
@azhavoro , could you please resolve the conflict?

File: cvat-core/src/server-proxy.ts [hunk 2]
```diff
@@ -509,13 +509,31 @@ async function healthCheck(maxRetries, checkPeriod, requestTimeout, progressCall
         timeout: requestTimeout,
     })
         .then((response) => response.data)
-        .catch((errorData) => {
-            if (maxRetries > 0) {
+        .catch((error) => {
+            let isHealthy = true;
+            let data;
+            if ('response' in error && typeof error.response.data === 'object') {
```
Comment (bsekachev) [inline cvat-core/src/server-proxy.ts:515]: Something like this should work:

```suggestion
            if (typeof error?.response?.data === 'object') {
```

File: cvat-core/src/server-proxy.ts [hunk 3]
```diff
@@ -509,13 +509,31 @@ async function healthCheck(maxRetries, checkPeriod, requestTimeout, progressCall
...
+                        checkName !== 'Cache backend: media' &&
+                        data[checkName] !== 'working') {
+                        isHealthy = false;
+                    }
+                }
+            } else {
+                isHealthy = false;
+            }
+
+            if (isHealthy === false && maxRetries > 0) {
```
Comment (bsekachev) [inline cvat-core/src/server-proxy.ts:528]:
```suggestion
            if (!isHealthy && maxRetries > 0) {
```

File: cvat-core/src/server-proxy.ts [hunk 4]
```diff
@@ -509,13 +509,31 @@ async function healthCheck(maxRetries, checkPeriod, requestTimeout, progressCall
...
+        .catch((error) => {
+            let isHealthy = true;
+            let data;
+            if ('response' in error && typeof error.response.data === 'object') {
+                data = error.response.data;
+                for (const checkName in data) {
+                    if (Object.prototype.hasOwnProperty.call(data, checkName) &&
+                        checkName !== 'Cache backend: media' &&
+                        data[checkName] !== 'working') {
+                        isHealthy = false;
```
Comment (bsekachev) [inline cvat-core/src/server-proxy.ts:521]: If it is a temporary workaround, let's keep a comment about it here

File: cvat-ui/src/messages.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2023 CVAT.ai Corp
```
Comment (bsekachev) [inline cvat-ui/src/messages.tsx:1]:
```suggestion
// Copyright (C) 2023 CVAT.ai Corporation
```

File: cvat-ui/src/messages.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        <a
+            target='_blank'
+            rel='noopener noreferrer'
+            href={UPGRADE_GUIDE_URL}
+        >
+            Upgrade Guide
+        </a>
+        .
+    </>
+);
```
Comment (bsekachev) [inline cvat-ui/src/messages.tsx:25]: If we want to override the message for app.cvat.ai, how would it work with this file?

Comment (azhavoro) [reply cvat-ui/src/messages.tsx:25]: The idea is to build UI image with a custom messages.tsx file... this is not a ideal solution but I can't find another approach that can be quickly implemented

**Review by nmanovic — APPROVED** [review]
From my perspective these changes are fine!

**Review by bsekachev — APPROVED** [review]

**Pull request merged**